MapboxNavigation/NavigationMapView.swift

```diff
@@ -427,36 +426,96 @@ open class NavigationMapView: MGLMapView {
-            if tapDistanceFromClosestRoute <= maximumTapDistanceToSelectAltRoute {
-                let selectedRoute = routes[routeIndex]
-                self.routes?.remove(at: routeIndex)
-                self.routes?.insert(selectedRoute, at: 0)
-                self.showRoutes(self.routes!)
-                navigationMapDelegate?.navigationMapView?(self, didTap: selectedRoute)
-            }
+        //lets filter to see which ones are under threshold
+        let candidates = closest.filter({
+            let coordinatePoint = self.convert($0.coordinate, toPointTo: self)
+            return coordinatePoint.distance(to: tapPoint) < waypointThreshold
+        })
+        
+        return (multipointRoutes, candidates)
+    }
+    
+    private func routes(closeTo tap: UITapGestureRecognizer) -> [Route]? {
+        guard let tapCoordinate = tap.coordinate(in: self), let tapPoint = tap.point else { return nil }
+        
+        //do we have routes? If so, filter routes with at least 2 coordinates.
+        guard let routes = routes?.filter({ $0.coordinates?.count ?? 0 > 1 }) else { return nil }
+        
+        //Sort routes by closest distance to tap gesture.
+        let closest = routes.sorted { (left, right) -> Bool in
+            
+            //existance has been assured through use of filter.
+            let leftLine = Polyline(left.coordinates!)
+            let rightLine = Polyline(right.coordinates!)
+            let leftDistance = leftLine.closestCoordinate(to: tapCoordinate)!.distance
+            let rightDistance = rightLine.closestCoordinate(to: tapCoordinate)!.distance
+            
+            return leftDistance < rightDistance
+        }
+        
+        //filter closest coordinates by which ones are under threshold.
+        let candidates = closest.filter {
+            let closestCoordinate = Polyline($0.coordinates!).closestCoordinate(to: tapCoordinate)!.coordinate
+            let closestPoint = self.convert(closestCoordinate, toPointTo: self)
+            
+            return closestPoint.distance(to: tapPoint) < maximumTapDistanceToSelectAltRoute
         }
+        return candidates
+    }
+    
+    private func select(route: Route) {
+        guard var routes = routes, let routeIndex = routes.index(where: {$0 == route}) else { return }
+        routes.remove(at: routeIndex)
+        routes.insert(route, at: 0)
+        self.routes = routes //because self.routes is a value type
+        self.showRoutes(routes)
+        navigationMapDelegate?.navigationMapView?(self, didTap: route)
+    }
+    
+    private func requestNew(route: Route, without waypoint: Waypoint) {
+        let options = route.routeOptions
+        guard let index = options.waypoints.index(of: waypoint) else { return }
+        options.waypoints.remove(at: index)
+        
+        removeRoutes()
+        removeWaypoints()
+        _ = Directions.shared.calculate(options) { (waypoints, routes, error) in
```

Comment:
It’s contrary to the MVC pattern for a view to kick off network requests. It should be the responsibility of the map view’s delegate (typically a view controller) to issue this request.

Reply:
Agreed. @bsudekum's idea of just putting this burden on the implementation side will suffice for now, but yeah, it would be nice if the `NavigationMapView` got its own controller that could provide default implementations for some of this behavior. It's time will come :)